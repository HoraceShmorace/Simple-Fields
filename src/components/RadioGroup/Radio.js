import { bool, func, number, oneOf, string } from 'prop-types'
import cx from 'classnames'
import styles from './Radio.module.css'

const Radio = ({ id, index, isDefaultSelected, label, onClick, type, value }) => {
  const uniqueId = `${id}_${index}`
  const className = cx([
    styles.root,
    styles[type],
    { [styles.checked] : isDefaultSelected}
  ])

  const handleClick = () => onClick(value)

  return (
    <li className={className}>
      <input
        defaultChecked={isDefaultSelected}
        id={uniqueId}
        name={id}
        onClick={handleClick}
        type='radio'
        defaultValue={value}
      />
      <button onClick={handleClick}>
        <div className={styles.indicator} />
      </button>
      <label htmlFor={uniqueId}>{label}</label>
    </li>
  )
}

Radio.propTypes = {
  id: string.isRequired,
  index: number.isRequired,
  isDefaultSelected: bool,
  label: string.isRequired,
  onClick: func,
  type: oneOf([
    'radio',
    'pill'
  ]),
  value: string
}

Radio.defaultProps = {
  isDefaultSelected: false,
  onClick: null,
  type: 'radio',
  value: ''
}

export default Radio
