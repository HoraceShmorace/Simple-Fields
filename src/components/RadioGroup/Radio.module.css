.root {
  align-items: center;
  flex-direction: row;
  margin-bottom: 1rem;
}

.root input[type=radio] {
  visibility: hidden;
  position: absolute;
}

.radio {
  display: flex;
}

.radio label {
  color: #666;
  cursor: pointer;
  font-size: 1rem;
  font-weight: bold;
  margin-left: 0.75rem;
}

.radio button {
  background-color: transparent;
  border: 3px solid #999;
  border-radius: 1rem;
  cursor: pointer;
  height: 1.5rem;
  padding: 3px;
  position: relative;
  width: 1.5rem;
}

.radio button div.indicator {
  background-color: #ccc;
  border-radius: 0.2rem;
  height: 90%;
  margin: auto;
  width: 90%;
}

.radio.checked button {
  background-color: #fff;
  border-color: #444;
}

.radio.checked button div.indicator {
  background-color: #ff9900;
  height: 99%;
  width: 99%;
}

.pill {
  display: inline-flex;
  background-color: #ccc;
  border: 2px solid #ddd;
  border-radius: 1rem;
  color: #666;
  cursor: pointer;
  margin: 0 4px;
}

.pill.checked {
  background-color: #336699;
  color: #eee;
}

.pill:first-child {
  margin-left: 0;
}

.pill:last-child {
  margin-right: 0;
}

.pill button {
  display: none;
}

.pill label {
  border-radius: 1rem;
  cursor: pointer;
  font-size: 1rem;
  font-weight: bold;
  height: 100%;
  padding: 8px 16px;
  width: 100%;
}
