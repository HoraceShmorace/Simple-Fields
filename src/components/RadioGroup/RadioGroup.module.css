.root ul {  
  list-style-type: none;
  margin: 0;
  padding-inline-start: 0;
  padding-left: 0;
}

.root ul li:last-child {
  margin-bottom: 0;
}

