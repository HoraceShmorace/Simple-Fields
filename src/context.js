import { createContext, useReducer } from 'react'
import { node } from 'prop-types'
const ADD_DATA_ACTION = 'ADD_DATA'
const RESET_DATA_ACTION = 'RESET_DATA'
const initialState = {}
const reducer = (state, { payload, type }) => {
  switch (type) {
    case ADD_DATA_ACTION:
      return {
        ...state,
        ...payload
      }
    case RESET_DATA_ACTION:
      return {}
    default:
      return state
  }
}

export const Context = createContext(initialState)

export const ContextProvider = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, initialState)
  const addData = payload => dispatch({ payload, type: ADD_DATA_ACTION })
  const resetData = () => dispatch({ payload: initialState, type: RESET_DATA_ACTION })

  return (
    <Context.Provider value={{ addData, resetData, state }}>
      {children}
    </Context.Provider>
  )
}

ContextProvider.propTypes = {
  children: node.isRequired
}
