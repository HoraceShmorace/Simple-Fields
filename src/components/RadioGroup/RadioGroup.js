import { useState } from 'react'
import { array, func, oneOf, string } from 'prop-types'
import Fieldset from '../Fieldset/Fieldset'
import Radio from './Radio'
import styles from './RadioGroup.module.css'

const RadioGroup = ({
  fields,
  id,
  label,
  onClick,
  type,
  value: defaultValue
}) => {
  const [selected, setSelected] = useState(defaultValue)
  const handleClick = (value) => {
    setSelected(value)
    if (typeof onClick === 'function') onClick({ [id]: value })
  }

  const RadioButtons = fields.map((props, index) => (
    <Radio
      {...props}
      key={`${id}_${index}`}
      id={id}
      index={index}
      isDefaultSelected={props.value === selected}
      onClick={handleClick}
      type={type}
    />
  ))

  return (
    <Fieldset className={styles.root} label={label}>
      <ul>
        {RadioButtons}
      </ul>
    </Fieldset>
  )
}

RadioGroup.propTypes = {
  fields: array,
  id: string.isRequired,
  onClick: func,
  type: oneOf([
    'radio',
    'pill'
  ]),
  value: string
}

RadioGroup.defaultProps = {
  fields: [],
  onClick: null,
  type: 'radio',
  value: ''
}

export default RadioGroup
