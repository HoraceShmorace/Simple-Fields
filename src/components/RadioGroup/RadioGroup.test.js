import { render } from 'enzyme'
import toJson from 'enzyme-to-json'
import RadioGroup from './RadioGroup'
import { ContextProvider } from '~src/context'
import mockData from '~mock/form.json'

const FIELD_TYPE_TEXT = 'radio'
const { fields, id } = mockData

describe('RadioGroup', () => {
  describe('when rendered', () => {
    const props = fields.find(({ type }) => type === FIELD_TYPE_TEXT)
    const component = render(
      <ContextProvider formId={id}>
        <RadioGroup {...props} />
      </ContextProvider>
    )

    it('should match snapshot', () => {
      expect(toJson(component)).toMatchSnapshot()
    })
  })
})
