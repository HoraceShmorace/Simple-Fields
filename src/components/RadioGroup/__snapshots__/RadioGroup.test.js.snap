// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioGroup when rendered should match snapshot 1`] = `
<ul
  class="fs-radio-group"
>
  <li>
    <input
      class="fs-radio"
      id="115390349_1"
      name="115390349"
      type="radio"
      value="Option1"
    />
    <label
      for="115390349_1"
    >
      Option1
    </label>
  </li>
  <li>
    <input
      class="fs-radio"
      id="115390349_2"
      name="115390349"
      type="radio"
      value="Option2"
    />
    <label
      for="115390349_2"
    >
      Option2
    </label>
  </li>
  <li>
    <input
      class="fs-radio"
      id="115390349_3"
      name="115390349"
      type="radio"
      value="Option3"
    />
    <label
      for="115390349_3"
    >
      Option3
    </label>
  </li>
</ul>
`;
